src/client/pydaos/pydaos_shim.c

```diff
@@ -115,58 +122,109 @@ __shim_handle__err_to_str(PyObject *self, PyObject *args)
  */
 
 static PyObject *
-cont_open(int ret, uuid_t puuid, uuid_t cuuid, int flags)
+cont_open(int ret, char *pool, char *cont, int flags)
 {
-	PyObject	*return_list;
-	daos_handle_t	 coh = {0};
-	daos_handle_t	 poh = {0};
-	int		 rc;
+	PyObject			*return_list;
+	struct open_handle		*hdl = NULL;
+	daos_handle_t			coh = {0};
+	daos_handle_t			poh = {0};
+	daos_prop_t			*prop = NULL;
+	struct daos_prop_entry		*entry;
+	struct daos_prop_co_roots	*roots;
+	int				rc;
 
 	if (ret != DER_SUCCESS) {
 		rc = ret;
 		goto out;
 	}
 
 	/** Connect to pool */
-	rc = daos_pool_connect(puuid, "daos_server", DAOS_PC_RW, &poh,
+	rc = daos_pool_connect(pool, NULL, DAOS_PC_RW, &poh,
 			       NULL, NULL);
 	if (rc)
 		goto out;
 
 	/** Open container */
-	rc = daos_cont_open(poh, cuuid, DAOS_COO_RW, &coh, NULL, NULL);
+	rc = daos_cont_open(poh, cont, DAOS_COO_RW, &coh, NULL, NULL);
 	if (rc)
-		daos_pool_disconnect(poh, NULL);
+		goto out;
+
+	/** Retrieve container properties via cont_query() */
+	prop = daos_prop_alloc(0);
+	if (prop == NULL) {
+		rc = -ENOMEM;
+		goto out;
+	}
+
+	rc = daos_cont_query(coh, NULL, prop, NULL);
+	if (rc)
+		goto out;
+
+	/** Verify that this is a python container */
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_LAYOUT_TYPE);
+	if (entry == NULL || entry->dpe_val != DAOS_PROP_CO_LAYOUT_PYTHON) {
+		D_ERROR("Container is not a python container\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	/** Fetch root object ID */
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_ROOTS);
+	if (entry == NULL) {
+		D_ERROR("Invalid entry in properties for root object ID\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+	roots = (struct daos_prop_co_roots *)entry->dpe_val_ptr;
+	if (roots->cr_oids[0].hi == 0 && roots->cr_oids[0].lo == 0) {
+		D_ERROR("Invalid root object ID in properties\n");
+		rc = -DER_INVAL;
+		goto out;
+	}
+
+	/** Track container open handle and associated attributes */
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL) {
+		D_ERROR("failed to allocate internal handle to open container\n");
+		rc = -DER_NOMEM;
+		goto out;
+	}
+	hdl->poh	= poh;
+	hdl->coh	= coh;
+	hdl->root	= roots->cr_oids[0];
+	/** Use flatkv option for root kv */
+	hdl->root.hi	|= (uint64_t)DAOS_OF_KV_FLAT << OID_FMT_FEAT_SHIFT;
+	hdl->alloc.lo	= 0;
+	hdl->alloc.hi	= MAX_OID_HI;
 out:
+	if (prop)
+		daos_prop_free(prop);
+	if (rc) {
+		if (daos_handle_is_valid(coh))
+			daos_cont_close(coh, NULL);
+		if (daos_handle_is_valid(poh))
+			daos_pool_disconnect(poh, NULL);
+	}
+
 	/* Populate return list */
-	return_list = PyList_New(3);
+	return_list = PyList_New(2);
```

Comment:
Should we have a NULL check after PyList_New?
According to https://docs.python.org/3/c-api/list.html
> PyObject* PyList_New(Py_ssize_t len)
Return value: New reference.
Return a new list of length len on success, or NULL on failure.

Reply:
Good point. It sounds like this is a broader issue than this patch and will look at this in another PR. If the allocation fails, then i am not sure how to report the error to the caller ...

Reply:
Unfortunately, it doesn't seem that PyList_New returns a proper error in all cases: https://github.com/python/cpython/blob/058fb35b57ca8c5063d16ec818e668b3babfea65/Objects/listobject.c#L140
But in some cases maybe an exception is raised for allocation failures?
Anyway, beyond this PR's scope